cli/runtime/config/config_factory.go

```diff
@@ -8,22 +8,161 @@ import (
+
+	// return the construct root node that contains both config.yaml with cfgItems and all of config-ng.yaml
+	return rootCfgNode, nil
 }
 
-type CfgOpts func(config *CfgOptions)
+// persistConfig write the updated node data to config.yaml and config-ng.yaml based on cfgItems
+func persistConfig(node *yaml.Node) error {
+	// check to persist multi file or to config-ng yaml
+	useUnifiedConfig, err := UseUnifiedConfig()
+	if err != nil {
+		useUnifiedConfig = false
+	}
+	// If useUnifiedConfig is set to true write to config-ng.yaml
+	if useUnifiedConfig {
+		return persistClientConfigNextGen(node)
+	}
+
+	// config node from config.yaml
+	cfgNode, err := getClientConfigNoLock()
+	if err != nil {
+		return err
+	}
+
+	// config items that goes to config.yaml
+	cfgItems := []string{
+		KeyAPIVersion,
+		KeyKind,
+		KeyMetadata,
+		KeyClientOptions,
+		KeyServers,
+		KeyCurrentServer,
+	}
 
-func WithCfgPath(path string) CfgOpts {
-	return func(config *CfgOptions) {
-		config.CfgPath = path
+	// Create a root cfg document node
+	rootCfgNode, err := newClientConfigNode()
+	if err != nil {
+		return errors.Wrap(err, "failed to create root client config node")
 	}
+
+	// Create a root cfg next gen document node
+	rootCfgNextGenNode, err := newClientConfigNode()
+	if err != nil {
+		return errors.Wrap(err, "failed to create root client config node")
+	}
+
+	// Loop through the change nodes and construct root cfg node and root cfg next gen node
+	for index, changeNode := range node.Content[0].Content {
+		if index%2 == 0 {
+			// If contains then add it to root config node for config.yaml
+			if collectionutils.Contains(cfgItems, changeNode.Value) {
+				rootCfgNode.Content[0].Content = append(rootCfgNode.Content[0].Content, node.Content[0].Content[index:index+2]...)
+			} else {
+				// Get the original node from config.yaml
+				changeNodeIndexOfCfgNode := nodeutils.GetNodeIndex(cfgNode.Content[0].Content, changeNode.Value)
+				// If exists then append to root cfg node for config yaml
+				if changeNodeIndexOfCfgNode != -1 {
+					rootCfgNode.Content[0].Content = append(rootCfgNode.Content[0].Content, cfgNode.Content[0].Content[changeNodeIndexOfCfgNode-1:changeNodeIndexOfCfgNode+1]...)
+				}
+				// Since the change node is not in root cfg node then Add this next gen node to root cfg next gen node for config-ng.yaml
+				rootCfgNextGenNode.Content[0].Content = append(rootCfgNextGenNode.Content[0].Content, node.Content[0].Content[index:index+2]...)
+			}
+		}
+	}
+
+	// Store the non nextGenItem config data to config.yaml
+	err = persistClientConfig(rootCfgNode)
+	if err != nil {
+		return err
+	}
+
+	// Store the nextGenItem config data to config-ng.yaml
+	err = persistClientConfigNextGen(rootCfgNextGenNode)
+	if err != nil {
+		return err
+	}
+
+	// Store the config data to legacy client config file/location
+	err = persistLegacyClientConfig(node)
```

Comment:
Thanks for updating the implementation. It is cleaner and less subject to change now.

Unless I am missing something, doesn't this line override what L145 is trying to do?

Reply:
legacy referred is .tanzu folder L145 is config.yaml.